File: C#/Array & String/443. String Compression.cs
/* 443. String Compression
Difficulty: Medium
Submission Link: https://leetcode.com/problems/string-compression/submissions/1030285250/ */

public class Solution {
    public int Compress(char[] chars) {
        int len = chars.Length;
        if (len == 0) return 0;

        char temp = chars[0];
        int count = 1;
        int res = 0;

        for (int i = 1; i <= len; i++) 
        {
            if (i < len && chars[i - 1] == chars[i]) count++;
            else 
            {
                chars[res] = temp;
                res += 1;

                string countStr = count.ToString();
                if (count > 1) 
                {
                    for (int j = 0; j < countStr.Length; j++) 
                        chars[res + j] = countStr[j];

                    res += countStr.Length;
                }

                if (i < len) temp = chars[i];
                count = 1;
            }
        }

        return res;
    }
}

File: C/Array & String/443. String Compression.c
/* 443. String Compression
Difficulty: Medium
Submission Link: https://leetcode.com/problems/string-compression/submissions/1030284831/
 */

int compress(char * const chars, const int charsLen){
	int newCharsLen = 0;

	for (int i = 0; i < charsLen; ){
		const char curChar = chars[i];
		i += 1;

		int curCharCnt = 1;
		while (i < charsLen && curChar == chars[i]){
			curCharCnt += 1;
			i += 1;
		}

		chars[newCharsLen] = curChar;
		newCharsLen += 1;

		if (curCharCnt > 1){
			char digits[10];
			int digitsLen = 0;
			while (curCharCnt > 0){
				digits[digitsLen] = curCharCnt % 10 + '0';
				digitsLen += 1;

				curCharCnt /= 10;
			}

			for (int j = digitsLen - 1; j >= 0; j -= 1){
				chars[newCharsLen] = digits[j];
				newCharsLen += 1;
			}
		}
	}

	return newCharsLen;
}